solr/core/src/java/org/apache/solr/jersey/JerseyAppHandlerCache.java

```diff
@@ -0,0 +1,98 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.jersey;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import org.apache.solr.core.ConfigSet;
+import org.apache.solr.core.SolrConfig;
+import org.apache.solr.core.SolrCore;
+import org.apache.solr.util.RefCounted;
+import org.glassfish.jersey.server.ApplicationHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Stores Jersey 'ApplicationHandler' instances by an ID or hash derived from their {@link
+ * ConfigSet}.
+ *
+ * <p>ApplicationHandler creation is expensive; caching these objects allows them to be shared by
+ * multiple cores with the same configuration.
+ */
+public class JerseyAppHandlerCache {
+
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private final Map<String, RefCounted<ApplicationHandler>> applicationByConfigSetId;
+
+  public JerseyAppHandlerCache() {
+    this.applicationByConfigSetId = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Return the 'ApplicationHandler' associated with the provided ID, creating it first if
+   * necessary.
+   *
+   * <p>This method is thread-safe by virtue of its delegation to {@link
+   * ConcurrentHashMap#computeIfAbsent(Object, Function)} internally.
+   *
+   * @param effectiveConfigSetId an ID to associate the ApplicationHandler with. Usually created via
+   *     {@link #generateIdForConfigSet(ConfigSet)}.
+   * @param createApplicationHandler a Supplier producing an ApplicationHandler
+   */
+  public RefCounted<ApplicationHandler> computeIfAbsent(
+      String effectiveConfigSetId, Supplier<ApplicationHandler> createApplicationHandler) {
+    final Function<String, RefCounted<ApplicationHandler>> wrapper =
+        s -> {
+          return new RefCounted<>(createApplicationHandler.get()) {
+            @Override
+            public void close() {
+              log.info(
+                  "Removing AppHandler from cache for 'effective configset' [{}]",
+                  effectiveConfigSetId);
+              applicationByConfigSetId.remove(effectiveConfigSetId);
```

Comment:
There is a race condition here if a core closes at the same time one is created with the same need of this handler.  The thread creating a core could fetch from the ConcurrentHashMap.computeIfAbsent but hasn't _quite_ called incref() yet.  Then the other thread calling `decref` will close this here and remove.  I'm not sure what the consequence is yet.

Reply:
OK.  Let me restate the race to make sure I understand:

1. coreA is created and puts its AppHandler in the cache (with refCount=1)
2. The Solr node receives a 'create' for coreB and a 'delete' for coreA come in in close proximity
3. coreB fetches the existing appHandler from the cache (but hasn't yet incref'd and returned)
4. coreA is deleted and calls its `decref()`
5. CoreA's call to decref() causes the AppHandler to be removed from the cache, even though coreB is about to get a reference to it.
6. coreB makes its incref() call and returns from JAHC.computeIfAbsent


It's definitely a race, good catch.

The most obvious impact here is that some other coreC created later would need to instantiate an AppHandler when it should've been able to reuse the one that coreB is already using.

That's the only impact that jumps out at me.  I was concerned at first that if coreB was deleted later on, its decref call would end up causing coreC's AppHandler to be removed from the cache.  But since ApplicationHandler doesn't implement `equals()` the cache-removal triggered when refCount==0 wouldn't actually do anything.

Reply:
I'll need to think about this a bit more.

I think we could trivially fix this by making JerseyAppHandlerCache.computeIfAbsent a synchronized method, but that seems like a bit of a blunt instrument that would do a good bit of over-locking.  Conversely we could create a different lock for each key (i.e. effective-configset ID) and do more fine-grained locking, but that feels complex for the risk and impact here.

Reply:
To close the loop here - this race condition ended up going away in the new PR by removing RefCounting altogether (and using a Caffeine cache implementation, as suggested by David there).